src/toolchain.rs

```diff
@@ -325,9 +326,31 @@ impl<'a> Toolchain<'a> {
                 .and_then(|s| s.parse().ok())
                 .unwrap_or(0);
             if recursion_count > env_var::RUST_RECURSION_COUNT_MAX - 1 {
+                let binary_lossy: String = binary.to_string_lossy().into();
+                if let (Some(component_name), Ok(Some(component_statuses)), Ok(Some(manifest))) = (
+                    component_for_bin(&binary_lossy),
+                    self.list_components(),
+                    self.get_manifest(),
+                ) {
+                    let component_status = component_statuses
+                        .iter()
+                        .find(|cs| cs.component.short_name(&manifest) == component_name)
+                        .expect(&format!(
+                            "component {} should be in the manifest",
+                            component_name
+                        ));
+                    if !component_status.available {
+                        return Err(ErrorKind::BinaryProvidedByUnavailableComponent(
+                            component_status.component.short_name(&manifest),
+                            binary_lossy,
+                            self.name.clone(),
+                        )
+                        .into());
+                    }
+                }
```

Comment:
One thing about this approach is it does slow this error path down. Previously, with the BinaryNotFound error, it would return the error instantly. Now, regardless of whether or not the component is available, it takes a moment to return the suggestion. Is there a better way I'm not seeing than looping over the component_statuses?

But since it's behind the path exists check it does not affect the happy path.

Reply:
I don't think it's a problem to have slowdown in the error reporting path, so I don't object to this.